cpp/src/communicator/metadata_payload_exchange/tag.cpp

```diff
@@ -0,0 +1,349 @@
+            (static_cast<std::uint64_t>(comm_->rank()) << 32) | next_message_id_++;
+
+        log.trace("send metadata to ", dst, " (message_id=", message_id, ")");
+        RAPIDSMPF_EXPECTS(dst != comm_->rank(), "sending message to ourselves");
+
+        std::size_t payload_size =
+            (message->data() != nullptr) ? message->data()->size : 0;
+
+        // Append metadata: [original_metadata][message_id][payload_size]
+        // Release the original metadata and append protocol fields to avoid copying
+        auto combined_metadata =
+            std::make_unique<std::vector<std::uint8_t>>(message->release_metadata());
+
+        // Reserve space for message_id and payload_size at the end
+        std::size_t original_size = combined_metadata->size();
+        combined_metadata->resize(
+            original_size + sizeof(std::uint64_t) + sizeof(std::size_t)
+        );
+
+        // Append message_id
+        std::memcpy(
+            combined_metadata->data() + original_size, &message_id, sizeof(std::uint64_t)
+        );
+
+        // Append payload_size
+        std::memcpy(
+            combined_metadata->data() + original_size + sizeof(std::uint64_t),
+            &payload_size,
+            sizeof(std::size_t)
+        );
+
+        fire_and_forget_.push_back(
+            comm_->send(std::move(combined_metadata), dst, metadata_tag_)
+        );
+
+        // Send data immediately after metadata (if any)
+        if (message->data() != nullptr) {
+            fire_and_forget_.push_back(
+                comm_->send(message->release_data(), dst, gpu_data_tag_)
+            );
+        }
+    }
+
+    statistics_->add_duration_stat("comms-interface-send-messages", Clock::now() - t0);
+}
+
+void TagMetadataPayloadExchange::progress() {
+    auto const t0 = Clock::now();
+
+    // Process all phases of the communication protocol
+    receive_metadata();
+    auto completed_metadata_only = setup_data_receives();
+    auto completed_data = complete_data_transfers();
+
+    // Store all completed messages
+    received_messages_.insert(
+        received_messages_.end(),
+        std::make_move_iterator(completed_metadata_only.begin()),
+        std::make_move_iterator(completed_metadata_only.end())
+    );
+    received_messages_.insert(
+        received_messages_.end(),
+        std::make_move_iterator(completed_data.begin()),
+        std::make_move_iterator(completed_data.end())
+    );
+
+    cleanup_completed_operations();
+
+    statistics_->add_duration_stat("comms-interface-progress", Clock::now() - t0);
+}
+
+std::vector<std::unique_ptr<MetadataPayloadExchange::Message>>
+TagMetadataPayloadExchange::recv() {
+    auto const t0 = Clock::now();
+
+    // Return all completed messages and clear the internal storage
+    auto messages = std::move(received_messages_);
+    received_messages_.clear();
+
+    statistics_->add_duration_stat("comms-interface-receive-messages", Clock::now() - t0);
+
+    return messages;
```

Comment:
I don't think this makes sense to time now, since it's just a pointer move, so the timing will be orders of magnitude longer than move.

I think this whole function should be (no need for the `clear()` because moving from a vector clears it):
```
return std::move(received_messages_);
```